library/core/src/ptr/const_ptr.rs

```diff
@@ -223,6 +232,77 @@ impl<T: ?Sized> *const T {
         self.with_addr(f(self.addr()))
     }
 
+    /// Gets the "address" portion of the pointer, and 'exposes' the "provenance" part for future
+    /// use in pointer-to-integer casts.
+    ///
+    /// This is equivalent to `self as usize`, which semantically discards *provenance* and
+    /// *address-space* information. Furthermore, this (like the `as` cast) has the implicit
+    /// side-effect of marking the provenance as 'exposed', so on platforms that support it you can
+    /// later call [`from_exposed_addr`][] to reconstitute the original pointer including its
+    /// provenance. (Reconstructing address space information, if required, is your responsibility.)
+    ///
+    /// Using this method means that code is *not* following Strict Provenance rules. Supporting
+    /// [`from_exposed_addr`][] complicates specification and reasoning and may not be supported by
+    /// tools that help you to stay conformant with the Rust memory model, so it is recommended to
+    /// use [`addr`][pointer::addr] wherever possible.
+    ///
+    /// On most platforms this will produce a value with the same bytes as the original pointer,
+    /// because all the bytes are dedicated to describing the address. Platforms which need to store
+    /// additional information in the pointer may not support this operation, since the 'expose'
+    /// side-effect which is required for [`from_exposed_addr`][] to work is typically not
+    /// available.
+    ///
+    /// This API and its claimed semantics are part of the Strict Provenance experiment, see the
+    /// [module documentation][crate::ptr] for details.
+    ///
+    /// [`from_exposed_addr`]: pointer::from_exposed_addr
+    #[must_use]
+    #[inline]
+    #[unstable(feature = "strict_provenance", issue = "95228")]
+    pub fn expose_addr(self) -> usize
+    where
+        T: Sized,
+    {
+        // FIXME(strict_provenance_magic): I am magic and should be a compiler intrinsic.
+        self as usize
+    }
+
+    /// Convert an address back to a pointer, picking up a previously 'exposed' provenance.
+    ///
+    /// This is equivalent to `addr as ptr`. The provenance of the returned pointer is that of *any*
+    /// pointer that was previously passed to [`expose_addr`][pointer::expose_addr]. If there is no
+    /// previously 'exposed' provenance that justifies the way this pointer will be used, the
+    /// program has undefined behavior. Note that there is no algorithm that decides which
+    /// provenance will be used. You can think of this as "guessing" the right provenance, and the
+    /// guess will be "maximally in your favor", in the sense that if there is any way to avoid
+    /// undefined behavior, then that is the guess that will be taken.
+    ///
+    /// On platforms with multiple address spaces, it is your responsibility to ensure the the
+    /// address makes sense in the address space that this pointer will be used with.
```

Comment:
FWIW I had no idea what to say about address spaces so if you have any suggestions @Gankra I'd be eager to hear them.

Reply:
I think it's fine to punt on for now.